### PR TITLE
Configurable mistake messages

### PR DESCRIPTION
**Pull request by Septem151**
# PR Description

This PR adds the ability to configure messages that display overhead and in the chat box for each mistake being tracked.

## Features

- Config entries for each room's mistakes
- Toggle for showing stackable "?" messages
- Toggle for showing stackable "?" messages based on total player raid mistakes, or current mistake

## Small Changes

- Added vscode generated files into `.gitignore`

## Screenshots

image

### CODE REVIEW

**Comment by QuestingPet** [comment]
Neat! I never prioritized this because I didn’t find it necessary nor did I know of anyone that wanted it. Judging by the fact that you went ahead and did it, it sounds like maybe you or some of your friends would like to customize this. Seems fair

Quick skim seems fine, though I’d prefer if you could keep the same code formatting (4 spaces for tabs) and bump the plug-in version. Will look more in depth sometime this week.

Also thanks for contributing! Hopefully the codebase wasn’t too hard to follow, though you probably had an easier time since it’s abstracted out and you didn’t have to mess with the detection logic itself :p

File: build.gradle [hunk 1]
```diff
@@ -24,7 +24,7 @@ dependencies {
 }
 
 group = 'com.toamistaketracker'
-version = '1.2'
+version = '1.3'
```
Comment (QuestingPet) [inline build.gradle:27]: Please add a change log entry in the README

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 2]
```diff
@@ -5,96 +5,101 @@
...
+    ZEBAK_BLOOD_CLOUD("Zebak Blood Cloud", (config) -> config.zebakBloodMessage(), "zebak-blood-cloud.png"),
+    ZEBAK_EARTHQUAKE("Zebak Scream", (config) -> config.zebakScreamMessage(), "zebak-scream.png"),
+    ZEBAK_WAVE("Zebak Wave", (config) -> config.zebakWaveMessage(), "zebak-wave.png"),
 
     // Scabaras
-    KEPHRI_BOMB("Kephri Bomb", "I'm exploding!", "kephri-bomb.png"),
-    KEPHRI_SWARM_HEAL("Kephri Swarm Heal", "The swarms are going in!", "kephri-swarm.png"),
-    KEPHRI_EGG_EXPLODE("Kephri Egg Explode", "I've been hatched!", ""),
+    KEPHRI_BOMB("Kephri Bomb", (config) -> config.kephriBombMessage(), "kephri-bomb.png"),
+    KEPHRI_SWARM_HEAL("Kephri Swarm Heal", (config) -> config.kephriSwarmMessage(), "kephri-swarm.png"),
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: This mistake is currently unused -- let's refrain from adding any of the unused ones in the config for now. Ideally we would just remove them, but I think I like keeping them around for now just to remember things I had in mind. I think for now it's just Kephri swarm and Kephri egg.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 3]
```diff
@@ -5,96 +5,101 @@
...
+    WARDENS_P2_WINDMILL("Wardens P2 Windmill", (config) -> config.wardensWindmillMessage(), ""),
+    WARDENS_P2_BOMBS("Wardens P2 Bombs", (config) -> config.wardensBombsMessage(), ""),
+    WARDENS_P2_BIND("Wardens P2 Bind", (config) -> config.wardensBindMessage(), "wardens-bind.png"),
+    WARDENS_P2_SPECIAL_PRAYER("Wardens P2 Special Prayer", (config) -> config.wardensPrayerMessage(),
+            "wardens-special-prayer.png"),
+    WARDENS_P3_EARTHQUAKE("Wardens P3 Slam", (config) -> config.wardensSlamMessage(), "wardens-earthquake.png"),
+    WARDENS_P3_AKKHA("Wardens P3 Akkha", (config) -> "", "wardens-akkha.png"),
+    WARDENS_P3_ZEBAK("Wardens P3 Zebak", (config) -> "", "wardens-zebak.png"),
+    WARDENS_P3_KEPHRI("Wardens P3 Kephri", (config) -> config.wardensKephriMessage(), "wardens-kephri.png"),
+    WARDENS_P3_BABA("Wardens P3 Ba-Ba", (config) -> config.wardensBabaMessage(), "wardens-baba.png"),
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: IMO these should be the same as their respective boss messages. So reuse the kephri explosion and baba falling boulder getters here.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 4]
```diff
@@ -5,96 +5,101 @@
...
+    WARDENS_P3_LIGHTNING("Wardens P3 Lightning", (config) -> "", "wardens-lightning.png"), // Too noisy
     ;
-
+
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: nit: Here and throughout, remove the extra indentation on an empty line

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 5]
```diff
@@ -5,96 +5,101 @@
...
+    WARDENS_P3_KEPHRI("Wardens P3 Kephri", (config) -> config.wardensKephriMessage(), "wardens-kephri.png"),
+    WARDENS_P3_BABA("Wardens P3 Ba-Ba", (config) -> config.wardensBabaMessage(), "wardens-baba.png"),
+    WARDENS_P3_LIGHTNING("Wardens P3 Lightning", (config) -> "", "wardens-lightning.png"), // Too noisy
     ;
-
+    
     private static final Set<ToaMistake> ROOM_DEATHS = EnumSet.of(
-            DEATH_HET, DEATH_CRONDIS, DEATH_SCABARAS, DEATH_APMEKEN, DEATH_WARDENS);
-
+    DEATH_HET, DEATH_CRONDIS, DEATH_SCABARAS, DEATH_APMEKEN, DEATH_WARDENS);
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: nit: let's keep the formatting the same. I can send you an exported formatter of what my setup is like.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 6]
```diff
@@ -133,31 +142,29 @@ public static ToaMistake toGroupedMistake(ToaMistake mistake) {
      *                     previously before this.
      * @return The mistake chat message to use for the raider
      */
-    public static String getChatMessageForMistakeCount(ToaMistake mistake, int mistakeCount) {
+    public static String getChatMessageForMistakeCount(ToaMistakeTrackerConfig config, ToaMistake mistake, int mistakeCount) {
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: Make sure to update the documentation for this method above

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 7]
```diff
@@ -133,31 +142,29 @@ public static ToaMistake toGroupedMistake(ToaMistake mistake) {
...
-            return getChatMessageForBabaSlam(mistake.getChatMessage(), mistakeCount);
+        if (config.stackQuestionMarks() && mistake.getChatMessage(config).equals("?")) {
+            return getStackingChatMessage(mistake.getChatMessage(config), mistakeCount);
+        } else if (mistake.getChatMessage(config).contains("|")) {
+            return getAlternatingChatMessage(mistake.getChatMessage(config), mistakeCount);
         }
-
-        return mistake.getChatMessage();
+        
+        return mistake.getChatMessage(config);
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: nit: Save `mistake.getChatMessage(config)` to a variable at the top of the method so it can be reused throughout, instead of re-computing it

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java [hunk 8]
```diff
@@ -23,9 +24,444 @@ default boolean showMistakesInChat() {
...
+    @ConfigItem(
+            keyName = "stackOnlySameMistakes",
+            name = "Stack Only Same Mistakes",
+            description = "When \"Stack Mistake Messages\" is toggled on, whether or not the amount of times the " +
+                    "mistake message stacks is based on the current mistake made or all raid mistakes.",
+            position = 3
+    )
+    default boolean stackOnlySameMistakes() {
+        return true;
+    }
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: > or all raid mistakes

Is this for the current raid? For example, if akkha orb mistakes so far is 2, but total raid mistakes (this raid) is 10, this would prepend 2 `?`s if this is true, and 10 `?`s if this is false, right? 

I guess my initial confusion was if this was for "all raid mistakes" ever as opposed to current raid. Wonder if there's a better way to get this across. Is this a feature you're particularly interested in, or just thought it could be neat to add?

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java [hunk 9]
```diff
@@ -23,9 +24,444 @@ default boolean showMistakesInChat() {
...
+    @ConfigItem(
+            keyName = "kephriEggMessage",
+            name = "Kephri Hatching Egg",
+            description = "Message to show when hit by a hatching egg in the Kephri room (separate multiple messages by \"|\").",
+            section = kephriMistakeSettings,
+            position = 2
+    )
+    default String kephriEggMessage() {
+        return "I've been hatched!";
+    }
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: As mentioned earlier, let's remove.

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java [hunk 10]
```diff
@@ -23,9 +24,444 @@ default boolean showMistakesInChat() {
...
+            position = 8
+    )
+    default String babaFallingBoulderMessage() {
+        return "It's raining!";
+    }
+
+    @ConfigItem(
+            keyName = "babaBananaMessage",
+            name = "Ba-Ba Banana",
+            description = "Message to show when hit by a falling boulder in the Ba-Ba room (separate multiple messages by \"|\").",
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: Update description

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java [hunk 11]
```diff
@@ -23,9 +24,444 @@ default boolean showMistakesInChat() {
...
+            description = "Settings for the messages shown on Het puzzle and Akkha room mistakes (separate multiple messages by \"|\").",
+            position = 5,
+            closedByDefault = true
+    )
+    String hetAndAkkhaMistakeSettings = "hetAndAkkhaMistakeSettings";
+
+    @ConfigItem(
+            keyName = "hetPuzzleBeamMessage",
+            name = "Het Puzzle Beam",
+            description = "Message to show when hit by the beam in the Het puzzle room (separate multiple messages by \"|\").",
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: Instead of repeating this suffix everywhere, maybe we can put it in at least just the section headers? Ideally we could put it somewhere once that's obvious but I can't think of a great place.

Thoughts?

**Review by QuestingPet — CHANGES_REQUESTED** [review]
Nice! Left a few comments below. Thanks!

Also do you have any example in-game screenshots of this working with the custom messages?

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java (hunk 8 above)
Comment (QuestingPet) [reply src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: This might actually work better as an enum with three options, instead of two separate booleans:

`Mistake Message Stacking`:
```
ALL_MISTAKES
SAME_MISTAKE_ONLY [default]
NONE
```

or something along those lines

File: src/main/java/com/toamistaketracker/ToaMistake.java (hunk 2 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistake.java]: Just to make sure I understand this correctly:
`KEPHRI_SWARM_HEAL` and `KEPHRI_EGG_EXPLODE` are the values currently not being used?

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java (hunk 8 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: ENUM for "Mistake Message Stacking" config option has been created, but unable to test right now due to game update. Changed PR to "Draft" status in the  meantime.

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java [hunk 12]
```diff
@@ -157,8 +157,10 @@ private void afterDetectAll() {
...
 
     private void addChatMessageForMistake(Raider raider, ToaMistake mistake) {
-        int mistakeCount = panel.getCurrentMistakeCountForPlayer(raider.getName(), mistake);
-        String msg = ToaMistake.getChatMessageForMistakeCount(mistake, mistakeCount);
+        int mistakeCount = config.mistakeMessageStacking() == StackingBehavior.SAME_MISTAKES_ONLY
+                ? panel.getCurrentMistakeCountForPlayer(raider.getName(), mistake)
+                : panel.getCurrentTotalMistakeCountForPlayer(raider.getName());
+        String msg = ToaMistake.getChatMessageForMistakeCount(config, mistake, mistakeCount);
 
         if (msg.isEmpty()) return;
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java]: Can we maybe trim the message that is returned here to a max size (maybe 40 chars)? Otherwise it's possible to just spam the chat with no limit on message size it seems. E.g.

image

(this keeps going for like 10x the size)

Please add a small comment as well explaining why we're limiting it

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 13]
```diff
@@ -125,39 +135,42 @@ public static ToaMistake toGroupedMistake(ToaMistake mistake) {
...
-    private static String getChatMessageForBabaSlam(String message, int mistakeCount) {
-        if (mistakeCount % 2 == 0) {
-            return message;
-        } else {
-            return ALTERNATE_BABA_SLAM_CHAT_MESSAGE;
-        }
+    private static String getAlternatingChatMessage(String message, int mistakeCount) {
+        String[] messageChoices = message.split("\\|");
+        String messageChoice = messageChoices[mistakeCount % messageChoices.length];
+        return messageChoice;
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: This leads to an exception if the message is just `|`:

```
java.lang.ArithmeticException: / by zero
```

Probably if the length of the split array is < 2 we maybe should just return empty string as it's not configured correctly, or even the default message for the mistake. Thoughts?

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 14]
```diff
@@ -125,39 +135,42 @@ public static ToaMistake toGroupedMistake(ToaMistake mistake) {
...
-     * mistake count of this mistake in the raid for the current raider, *previous* to this mistake happening.
+     * Retrieve the chat message for the given mistake, considering special cases given the config
+     * settings about whether to stack question marks and the current mistake count of either
+     * this specific mistake in the raid for the current raider, *or* the total current raid mistake
+     * count for the current raider.
      *
      * @param mistake      The mistake
-     * @param mistakeCount The current mistake count of this mistake for the current raider in this current raid,
-     *                     previously before this.
+     * @param mistakeCount The current mistake count of this mistake *or* the current raid mistake count for the
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java:171]: nit: Add the new param here

**Review by QuestingPet — CHANGES_REQUESTED** [review]
Couple more comments for issues I found while testing

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java (hunk 12 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java]: Ah, good catch. I'll have to ask Adam/runelite discord on how to limit the size of configuration options entered. Either that, or manually truncating the text.

File: src/main/java/com/toamistaketracker/ToaMistake.java (hunk 13 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistake.java]: Change funnels the user-inputted message through a stream which splits the input by "|", then filters all empty messages to prevent things like "|||f|||" from looking like: `["", "", "f", "",  ""]` and instead becoming a single choice of `["f"]`. This still allows things like "f| " which would become: `["f", " "]` since a space message is not empty.

Additionally, if the input message splits into an empty array (aka the input is just "|" or "|||||" etc) then the default message is shown instead.

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java (hunk 12 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistakeTrackerPlugin.java]: Mistake message truncated to a max of 40 characters now, including split messages.
Each split message is allowed to be up to 40 characters, so for example:
"Nam quis nulla. Integer malesuada. In in enim|a arcu imperdiet malesuada. Sed vel lectus. Don" will display the messages:
- "Nam quis nulla. Integer malesuada. In in"
- "a arcu imperdiet malesuada. Sed vel lect"

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 15]
```diff
@@ -6,71 +6,87 @@
...
 
 public enum ToaMistake {
     // Deaths
-    DEATH("Death", (config) -> "", "death.png"), // Chat message handled in deaths below
-    DEATH_HET("Path of Het Death", (config) -> config.deathMessage(), "death-akkha.png"),
-    DEATH_CRONDIS("Path of Crondis Death", (config) -> config.deathMessage(), "death-zebak.png"),
-    DEATH_SCABARAS("Path of Scabaras Death", (config) -> config.deathMessage(), "death-kephri.png"),
-    DEATH_APMEKEN("Path of Apmeken Death", (config) -> config.deathMessage(), "death-baba.png"),
-    DEATH_WARDENS("Wardens Death", (config) -> config.deathMessage(), "death-wardens.png"),
+    DEATH("Death", (config) -> "", "I'm planking!", "death.png"), // Chat message handled in deaths below
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: Let's keep the default here as empty string. I misremembered that if we actually have this spit out a non-empty message (which it could if we have to fallback to the default), then it will display both messages when you die (the top-level one and the room-level one below).

Probably shouldn't work that way in the first place, but it's not worth changing now IMO

To get around the duplication issue, and the fact that you need to choose one for the config, let's create a static constant at the top `DEFAULT_DEATH_MESSAGE` and use for all the room deaths (and the death config). If we ever want to have different death messages per room (not really planned as of now) then we can worry about changing that then.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 16]
```diff
@@ -6,71 +6,87 @@
...
+    DEATH_HET("Path of Het Death", (config) -> config.deathMessage(), "I'm planking!", "death-akkha.png"),
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: nit: We could replace the lambda with the method reference (here and everywhere else): 
```
ToaMistakeTrackerConfig::deathMessage
```
Doesn't matter much either way, but it's a bit easier to read and know exactly which method and from which class this is coming from.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 17]
```diff
@@ -84,10 +100,11 @@ public enum ToaMistake {
     private static final String FALLBACK_IMAGE_PATH = "death.png";
 
     private static final int MAX_STACKING_CHAT_MESSAGE_LENGTH = 10;
+    public static final int MAX_MESSAGE_LENGTH = 40;
 
     @Getter
     @NonNull
-    private final String mistakeName;
+    private final String mistakeName, defaultMessage;
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: Not a fan of multi-declaration on a single line. Please put each field on its own.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 18]
```diff
@@ -152,11 +171,13 @@ public static String getChatMessageForMistakeCount(ToaMistakeTrackerConfig confi
...
-        } else if (mistakeMessage.contains("|")) {
-            return getAlternatingChatMessage(mistakeMessage, mistakeCount);
+        }
+        if (mistakeMessage.contains("|")) {
+            return getAlternatingChatMessage(mistakeMessage, mistake.getDefaultMessage(), mistakeCount);
         }
 
-        return mistakeMessage;
+        // Truncate message length to prevent overly-spammy messages taking up too much screen space
+        return truncateMessage(mistakeMessage);
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: I think we should do this a single time in the calling method `addChatMessageForMistake`, potentially right before or after the empty string check.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 19]
```diff
@@ -168,9 +189,21 @@ private static String getStackingChatMessage(String message, int mistakeCount) {
...
+    private static String getAlternatingChatMessage(String message, String defaultMessage, int mistakeCount) {
+        String[] messageChoices = Arrays.stream(message.split("\\|")).filter(msg -> !msg.isEmpty())
+                .toArray(String[]::new);
+        if (messageChoices.length < 1) {
+            return defaultMessage;
+        }
         String messageChoice = messageChoices[mistakeCount % messageChoices.length];
-        return messageChoice;
+        // Truncate message length to prevent overly-spammy messages taking up too much screen space
+        return truncateMessage(messageChoice);
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: See comment above. These helper methods don't need to worry about the truncation logic.

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 20]
```diff
@@ -168,9 +189,21 @@ private static String getStackingChatMessage(String message, int mistakeCount) {
         return sb.toString();
     }
 
-    private static String getAlternatingChatMessage(String message, int mistakeCount) {
-        String[] messageChoices = message.split("\\|");
+    private static String getAlternatingChatMessage(String message, String defaultMessage, int mistakeCount) {
+        String[] messageChoices = Arrays.stream(message.split("\\|")).filter(msg -> !msg.isEmpty())
+                .toArray(String[]::new);
+        if (messageChoices.length < 1) {
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: nit: The number here should be the one that is the most logical -- in our case, we want to return the default if there are no choices (empty, or 0 length). Let's change to `messageChoices.length == 0` since it's a bit easier to read, and 0 is more recognized as a "magic number" than 1

File: src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java [hunk 21]
```diff
@@ -383,7 +383,7 @@ default String wardensWindmillMessage() {
             position = 3
     )
     default String wardensBombsMessage() {
-        return "I'm getting bombed!";
+        return ToaMistake.WARDENS_P3_KEPHRI.getDefaultMessage();
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistakeTrackerConfig.java]: This should be the p2 bombs message, not the kephri message

File: src/main/java/com/toamistaketracker/ToaMistake.java [hunk 22]
```diff
@@ -6,71 +6,87 @@
...
-    WARDENS_P3_BABA("Wardens P3 Ba-Ba", (config) -> config.babaFallingBoulderMessage(), "wardens-baba.png"),
-    WARDENS_P3_LIGHTNING("Wardens P3 Lightning", (config) -> "", "wardens-lightning.png"), // Too noisy
+            "What even was that attack?", "wardens-special-prayer.png"),
+    WARDENS_P3_EARTHQUAKE("Wardens P3 Slam", (config) -> config.wardensSlamMessage(), "I'm tripping!",
+            "wardens-earthquake.png"),
+    WARDENS_P3_AKKHA("Wardens P3 Akkha", (config) -> "", "", "wardens-akkha.png"),
+    WARDENS_P3_ZEBAK("Wardens P3 Zebak", (config) -> "", "", "wardens-zebak.png"),
+    WARDENS_P3_KEPHRI("Wardens P3 Kephri", (config) -> config.kephriBombMessage(), "I'm exploding!",
+            "wardens-kephri.png"),
+    WARDENS_P3_BABA("Wardens P3 Ba-Ba", (config) -> config.babaFallingBoulderMessage(), "It's raining!",
```
Comment (QuestingPet) [inline src/main/java/com/toamistaketracker/ToaMistake.java]: These  defaults should also probably reference their above counterparts:

`"I'm exploding!"` -> `KEPHRI_BOMB.getDefaultMessage()` and same for the baba one

**Review by QuestingPet — CHANGES_REQUESTED** [review]

File: src/main/java/com/toamistaketracker/ToaMistake.java (hunk 22 above)
Comment (Septem151) [reply src/main/java/com/toamistaketracker/ToaMistake.java]: Dow.  Must've missed those
